### ISSUE TITLE
[Question]: No such property: params for class: com....

### ISSUE THREAD

**Issue by js-timbirkett** [jenkinsci/JenkinsPipelineUnit#324]
Hi 👋 

I have been lumbered with a bastardised mess of a pipeline that has grown from Java developers writing all of the pipeline code under the `/src` - literally, no bite-sized steps under `/vars`, not even a `/vars` directory 🤦

Inside the code, they reference the params object (parameters passed in from the Jenkins job) which works at runtime. Whilst trying to get some basic testing in place to start refactoring this thing I get the error:
```
groovy.lang.MissingPropertyException: No such property: params for class: com.xxxxxxx.QualysNightlyPipeline
```

I have a TestHelper that sets up the environment and registers allowed methods. It has this in it:
```
         /**
         * Job params - may need to override in specific tests
         */
        binding.setVariable('params', [:])
```

and the test itself has:
```
   given:
        def params = binding.getVariable('params')
        params['reposToScan'] = 'test-repo'
```

I guess my questions are:
- Is this because the groovy code is being evaluated/validated before it's getting to the actual test that might have the appropriate bindings set?
- Is it because groovy is trying to look up the value from the Class rather than from the binding?
- Anyone else encountered problems like this? Did you solve it?

I know the correct answer would probably be: refactor the code out of `/src` into `/vars` but I'd like to have some assurances in place first so that the development team are more confident in doing the refactor (which should probably start with `rm -rf *`.

**Comment by js-timbirkett** [comment]
Seriously... What have you done...? 🙈 
<img width="423" alt="Screenshot 2020-12-14 at 11 49 53 AM" src="https://user-images.githubusercontent.com/57101177/102078370-bc9a0800-3e02-11eb-82fc-6a5615e16224.png">

**Comment by nre-ableton** [comment]
tl;dr: You probably need a script context, which you should pass from `vars` to the class in `src`. However, I would advise against the refactoring you suggest, for reasons that I will explain below.

---

I would advise _against_ refactoring everything from `src` to `vars`. I believe that pipeline libraries are actually **much** easier to test when everything is under `src`, and `vars` just offers simple passthru functions. The reason this is the case is because this structure makes it much easier to mock pipeline steps, and also it really helps to manage dependencies between pipeline classes, should you have them.

First of all, the reason that you are (probably) seeing this error is because the classes under `src` don't have a connection to the pipeline itself. They are just regular Groovy classes and don't know about pipeline steps or special fields such as `params`.

When Jenkins executes a `Jenkinsfile`, it transforms it into a `WorkflowScript` class. This class knows about all the pipeline steps like `sh` and such, but also has instances of singletons for all files in any pipeline `vars` directories that were loaded from various libraries. These classes are also transformed by Jenkins to objects, which subclass the `CpsScript` class. Thus, they can also call pipeline steps, reference fields like `params`, and so on.

Anyways, my usual pattern for developing a pipeline library looks something like this:

```groovy
// src/com/whatever/Foo.groovy
package com.whatever

class Foo implements Serializable {
  Object script = null

  void bar() {
    assert script
    script.echo 'Hello from bar!'
  }

  bool myVarCorrect() {
    return script.params.MYVAR == 'secret'
  }
}
```

```groovy
// vars/foo.groovy
import com.whatever.Foo

void bar() {
  new Foo(script: this).bar()
}

bool myVarCorrect() {
  return new Foo(script: this).myVarCorrect()
}
```

```groovy
// test/com/whatever/FooTest.groovy
package com.whatever

import static org.junit.Assert.assertTrue

import com.lesfurets.jenkins.unit.BasePipelineTest
import org.junit.Before
import org.junit.Test

class FooTest extends BasePipelineTest {
  Object script = null

  @Override
  @Before
  void setUp() {
    super.setUp()
    this.script = loadScript('test/resources/EmptyPipeline.groovy')  // a file which has only "return this"
  }

  @Test
  void bar() {
    Foo foo = new Foo(script: script)
    foo.bar()
    // assert, etc.
  }

  @Test
  void myVarCorrect() {
    script.params = [MYVAR: 'secret']
    assertTrue(new Foo(script: script).myVarCorrect())
  }
}
```

...and finally, here's how the whole thing looks in a client Jenkinsfile:

```groovy
// Jenkinsfile
import 'foo@x.y.z'

node('master') {
  stage('bar') {
    if (foo.myVarCorrect()) {
      foo.bar()
    }
  }
}
```

So given this pattern, you need to inject the classes within `src` with a script context, usually upon construction. Then any pipeline step or variable (including `params`) will need to use the script context to access it.

In my experience, this pattern makes it much easier to develop and test pipelines, especially because if your libraries have lots of classes which reference each other, you get more guarantees that the objects are what you say they are. Also, if you are using an IDE (such as IntelliJ), you get improved type checking, refactoring etc.

If you want to see some open-source examples of such pipeline libraries, here's a few that my company has released which follow this pattern:

- https://github.com/Ableton/postgres-pipeline-utils/
- https://github.com/Ableton/python-pipeline-utils/

**Comment by js-timbirkett** [comment]
🤗 - thanks for taking the time to give such an in-depth reply and enlighten me to a few things.

I would agree that that is a nice approach to utility/pipeline helpers. But is a long way from the current state of play. I especially like the comment re: IDE syntax highlighting, type checking and refactoring capabilities. That makes a lot of sense.

Maybe a bit of careful refactoring on low value/risk things will be a prerequisite to retro-fitting working tests.

**Comment by js-timbirkett** [comment]
Currently, the script I'm trying to test is a Jenkinsfile. which looks like:
```
@Library('ci-toolkit') _
import com.xxxxxxx.ci.Pipelines

Pipelines pipelines = new Pipelines()

pipelines.qualysNightlyPipeline()
```
`qualysNightlyPipeline()` is where the reference to `params` is made inside a file names `QualysNightlyPipeline.groovy` there is no explicit class definition in the file so I'm assuming that groovy wraps the methods within in a class (whatever the file is called) automatically. 🤔

**Comment by nre-ableton** [comment]
I think you can probably do:

```groovy
@Library('ci-toolkit') _
import com.xxxxxxx.ci.Pipelines

Pipelines pipelines = new Pipelines(script: this)

pipelines.qualysNightlyPipeline()
```

And within the `Pipelines` class, declare an `Object script` context as detailed above, and use it to reference `script.params`.
